deps/media-playback/media-playback/media.c

```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
-		is_active = m->active;
-		pthread_mutex_unlock(&m->mutex);
+/*Executes after leaving a state. Keep in mind that this won't execute
+  until AFTER the entry method of the new state. Therefore, if cleanup
+  is needed before changing states, put that logic in the entry
+  function of the new state.*/
+static inline void mp_media_change_state(mp_media_t *m,
+					 enum mp_media_state oldstate)
+{
+	switch (oldstate) {
+	case mstate_uninit:
 
-		if (!is_active) {
-			if (os_sem_wait(m->sem) < 0)
-				return false;
-		} else {
-			timeout = mp_media_sleepto(m);
-		}
+		break;
+	case mstate_playing:
 
-		pthread_mutex_lock(&m->mutex);
+		break;
+	case mstate_paused:
+		/*reset time stamp to prevent fast forward*/
+		reset_ts(m);
+		break;
+	case mstate_resume:
 
-		reset = m->reset;
-		kill = m->kill;
-		m->reset = false;
-		m->kill = false;
+		break;
+	case mstate_seeking:
 
-		pause = m->pause;
-		seek_pos = m->seek_pos;
-		seek = m->seek;
-		reset_time = m->reset_ts;
-		m->seek = false;
-		m->reset_ts = false;
+		break;
+	case mstate_resetting:
 
-		pthread_mutex_unlock(&m->mutex);
+		break;
+	case mstate_setloop:
 
-		if (kill) {
-			break;
-		}
-		if (reset) {
-			mp_media_reset(m);
-			continue;
-		}
+		break;
+	case mstate_setreconnect:
 
-		if (seek) {
-			m->seek_next_ts = true;
-			seek_to(m, seek_pos);
-			continue;
-		}
+		break;
+	case mstate_stopping:
 
-		if (reset_time) {
-			reset_ts(m);
-			continue;
-		}
+		break;
+	case mstate_stopped:
 
-		if (pause)
-			continue;
+		break;
+	case mstate_killing:
 
-		/* frames are ready */
-		if (is_active && !timeout) {
-			if (m->has_video)
-				mp_media_next_video(m, false);
-			if (m->has_audio)
-				mp_media_next_audio(m);
+		break;
+	}
+}
 
-			if (!mp_media_prepare_frames(m))
-				return false;
-			if (mp_media_eof(m))
-				continue;
+/*Executes the top message in the queue.*/
+static bool mp_media_pop_message(mp_media_t *m)
+{
+	bool result = true;
+
+	mp_media_queue_node_t *front = m->mp_queue.first;
+	/*Only execute if a state change actually occurs. If an
+	  action needs to be called within the state, add a new
+	  state that executes the action then moves back.*/
+	if (m->mp_state != front->message.state) {
+		result = front->message.state_handler(m);
+	}
+
+	pthread_mutex_lock(&m->mutex);
+
+	if (m->mp_queue.first->next != NULL)
+		m->mp_queue.first = front->next;
+	else {
+		m->mp_queue.first = NULL;
+		m->mp_queue.last = NULL;
+	}
+	m->mp_queue.flags -= front->message.state;
+	free(front);
+
+	pthread_mutex_unlock(&m->mutex);
+
+	return result;
+}
+
+static bool mp_media_push_message(mp_media_t *m, enum mp_media_state state)
+{
+	/*To prevent the queue from growing infinitely, and to
+	  prevent needless repetition, use a bit flag.
+	  Queue should only contain any given message once.*/
+	if (m->mp_queue.flags & state)
+		return false;
+
+	mp_media_queue_node_t *newnode = malloc(sizeof(mp_media_queue_node_t));
+	mp_media_state_message_t *newmessage;
```

Comment:
I'd suggest using either `new_message` or `message`.  Not sure the point of using "new" since there is no other variable in this scope named `message`.